src/common/checksum.c

```diff
@@ -486,3 +502,39 @@ checksum_sgl_cb(uint8_t *buf, size_t len, void *args)
 
 	return daos_csummer_update(obj, buf, len);
 }
+
+int
+daos_csum_check_sgl(const daos_iod_t *iod, d_sg_list_t *sgl)
+{
+	struct daos_csummer	*csummer;
+	daos_csum_buf_t		*csum_bufs;
+	daos_csum_buf_t		*csum = iod->iod_csums;
+	int			 rc;
+
+	if (!dcb_is_valid(csum))
+		return 0;
+
+	rc = daos_csummer_type_init(&csummer,
+				    (enum DAOS_CSUM_TYPE) csum->cs_type,
+				    csum->cs_chunksize);
+	if (rc != 0) {
+		D_ERROR("Issue initializing csummer. "
+			"Unable to check data. Returning success so checksum "
+			"doesn't kill IO.");
+
+		return 0;
+	}
+
+	daos_csummer_calc_csum(csummer, sgl, iod->iod_size,
+			       iod->iod_recxs, iod->iod_nr, &csum_bufs);
+
+	if (!daos_csummer_compare(csummer, csum_bufs, csum)) {
+		D_WARN("Corruption found!!\n");
+		rc = -DER_IO;
```

Comment:
Is this the right error message for this?  Do we need a new one?  DER_IO is pretty generic and I wouldn't expect the client to know that it means it needs to try another target

Reply:
Good point. Maybe we add DER_DATA_CORRUPTION?